assets/src/stories-editor/components/right-click-menu/index.js

```diff
@@ -0,0 +1,322 @@
+/* global navigator */
+/**
+ * External dependencies
+ */
+import { castArray } from 'lodash';
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { cloneBlock, pasteHandler, serialize } from '@wordpress/blocks';
+import { useEffect, useState, useRef } from '@wordpress/element';
+import {
+	MenuGroup,
+	MenuItem,
+	NavigableMenu,
+	Popover,
+} from '@wordpress/components';
+import { withDispatch, withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import './edit.css';
+import {
+	copyTextToClipBoard,
+	ensureAllowedBlocksOnPaste,
+} from '../../helpers';
+
+const POPOVER_PROPS = {
+	className: 'amp-story-right-click-menu__popover block-editor-block-settings-menu__popover editor-block-settings-menu__popover',
+	position: 'bottom left',
+};
+
+function useOutsideClickChecker( ref, onClose ) {
+	/**
+	 * Close the Popover if outside click was detected.
+	 *
+	 * @param {Object} event Click event.
+	 */
+	function handleClickOutside( event ) {
+		if ( ref.current && ! ref.current.contains( event.target ) ) {
+			onClose();
+		}
+	}
+
+	useEffect( () => {
+		// Handle click outside only if the the menu has been added.
+		if ( ref.current && ref.current.innerHTML ) {
+			document.addEventListener( 'mousedown', handleClickOutside );
+		}
+		return () => {
+			// Unbind when cleaning up.
+			document.removeEventListener( 'mousedown', handleClickOutside );
+		};
+	} );
+}
+
+const RightClickMenu = ( props ) => {
+	const {
+		clientIds,
+		clientX,
+		clientY,
+		copyBlock,
+		cutBlock,
+		getBlock,
+		getCopiedMarkup,
+		removeBlock,
+		duplicateBlock,
+		pasteBlock,
+	} = props;
+	const [ isOpen, setIsOpen ] = useState( true );
+
+	useEffect( () => {
+		setIsOpen( true );
+	}, [ clientIds, clientX, clientY ] );
+
+	const blockClientIds = castArray( clientIds );
+	const firstBlockClientId = blockClientIds[ 0 ];
+	const block = getBlock( firstBlockClientId );
+	const isPageBlock = block ? 'amp/amp-story-page' === block.name : false;
+
+	const onClose = () => {
+		setIsOpen( false );
+	};
+
+	const containerRef = useRef( null );
+	useOutsideClickChecker( containerRef, onClose );
+
+	const onCopy = () => {
+		onClose();
+		copyBlock( firstBlockClientId );
+	};
+
+	const onCut = () => {
+		onClose();
+		cutBlock( firstBlockClientId );
+	};
+
+	const onRemove = () => {
+		onClose();
+		removeBlock( firstBlockClientId );
+	};
+
+	const onDuplicate = () => {
+		onClose();
+		duplicateBlock( firstBlockClientId );
+	};
+
+	const onPaste = () => {
+		pasteBlock( firstBlockClientId );
```

Comment:
I assume there's an excellent reason for the changed ordering here. Could you add a comment explaining it, so someone doesn't come along and changes this in some attempt of generalising these callbacks?

Reply:
Ha! Re-organized this, there was no reason for changing the order here, except for "not seeing the forest through the trees".